File: res/logo.lua

-- Constants
local pix_size = 320

local color0 = { 0, 0, 0.8 }
local color1 = { 0, 0.8, 0 }
local color2 = { 0.8, 0, 0 }
local table_size_x = 3.0
local table_size_y = 2.1
local wall_width  = 0.070
local wall_height = 0.050
local start_c = 0.9
local robot_c = 0.20
local dot_r = 0.1
local cross_r = dot_r*1.2
-- Robot position
local rx, ry, rz, ra = 0.9, -0.6, robot_c, 0.92*math.pi

-- Config
config.bg_color = { 0.8, 0.8, 0.8 }

config.draw_epsilon = 0.0005
config.draw_div = 20

config.perspective_fov = 45.0
config.perspective_near = 0.1
config.perspective_far = 300.0

config.screen_x = pix_size
config.screen_y = pix_size
config.fullscreen = false
config.fps = 60
config.antialias = 6

physics = Physics()
display = Display()


require('modules/colors')

-- Camera
display:set_camera_mode(display.CAM.FIXED)
display:set_camera_target({ x=0.1, y=-0.2, z=0 })
display:set_camera_eye({ r=5, theta=0.6, phi=-1.0 })


---- Field

-- Return position on robot's axis
function raxis_pos(r)
  return rx+r*math.cos(ra), ry+r*math.sin(ra), rz
end


-- Robot
o = OSimple()
o:set_shape( Shape:box(robot_c,robot_c,robot_c) )
o:add_to_world()
o:set_color({1,1,0})
o:set_pos( rx, ry, rz+0.01 )
o:set_rot( ra, 0, 0 )

-- Dots
sh = Shape:sphere(dot_r)
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_color(colors.white)
o:set_pos( raxis_pos(0.50) )
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_color(colors.white)
o:set_pos( raxis_pos(0.90) )

-- Cross
--sh = Shape:capsuleY(dot_r,3*dot_r)
sh = Shape:box(cross_r,3*cross_r,cross_r)
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_color(colors.white)
o:set_pos( raxis_pos(1.50) )
o:set_rot( ra+math.pi/4, 0, 0 )
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_color(colors.white)
o:set_pos( raxis_pos(1.50) )
o:set_rot( ra+3*math.pi/4, 0, 0 )


-- Ground
trace("  ground")

o = OGround(color0, color1, color2)
o:add_to_world()

sh = Shape:box(start_c/2,start_c/2,config.draw_epsilon*2)
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_pos( -(table_size_x-start_c)/2, (table_size_y-start_c)/2, 0 )
o:set_color(color1)
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_pos( (table_size_x-start_c)/2, (table_size_y-start_c)/2, 0 )
o:set_color(color2)


-- Walls (N, S, E, W)
trace("  walls")

sh = Shape:box(table_size_x/2+wall_width, wall_width/2, wall_height/2)
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_pos(0, table_size_y/2+wall_width/2, wall_height/2)
o:set_color(colors.black)
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_pos(0, -table_size_y/2-wall_width/2, wall_height/2)
o:set_color(colors.black)

sh = Shape:box(wall_width/2, table_size_y/2+wall_width, wall_height/2)
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_pos(table_size_x/2+wall_width/2, 0, wall_height/2)
o:set_color(colors.black)
o = OSimple()
o:set_shape( sh )
o:add_to_world()
o:set_pos(-table_size_x/2-wall_width/2, 0, wall_height/2)
o:set_color(colors.black)

File: res/logo.py
#!/usr/bin/env python

import sys, os
sys.path.append('.')  # allow to run from project root
import math
from simulotter import *


# Logo configuration constants

color0 = Color4(0,0,0.8)
color1 = Color4(0,0.8,0)
color2 = Color4(0.8,0,0)

table_size = vec2(3.0, 2.1)
wall_width  = 0.070
wall_height = 0.050
start_c = 0.9
robot_c = 0.20
dot_r = 0.1
cross_r = dot_r * 1.2

robot_tr = trans( quat(0, 0, 0.92*math.pi), vec3(0.9, -0.6, robot_c) )


def build_logo(ph):
  """Add logo objects to a Physics instance."""

  # Robot
  sh = ShBox(vec3(robot_c,robot_c,robot_c))
  robot = OSimple(sh)
  robot.color = Color4(1.,1.,0.)
  robot.trans = robot_tr
  robot.addToWorld(ph)

  # Dots
  sh = ShSphere(dot_r)
  for x in (0.50, 0.90):
    o = OSimple(sh)
    o.color = Color4.white()
    o.pos = robot_tr * vec3(x, 0, 0)
    o.addToWorld(ph)

  # Cross
  sh = ShBox(vec3(cross_r,3*cross_r,cross_r))
  for a in (math.pi/4, 3*math.pi/4):
    o = OSimple(sh)
    o.color = Color4.white()
    o.pos = robot_tr * vec3(1.50, 0, 0)
    o.rot = robot_tr.basis * quat(0, 0, a)
    o.addToWorld(ph)

  # Ground
  o = OGroundSquareStart(color0, color1, color2)
  o.addToWorld(ph)

  sh = ShBox(vec3(start_c*0.5, start_c*0.5, Display.draw_epsilon*2))
  v = 0.5 * ( table_size - vec2(start_c, start_c) )
  o = OSimple(sh)
  o.color = color1
  o.pos = vec2(-v.x, v.y)
  o.addToWorld(ph)
  o = OSimple(sh)
  o.color = color2
  o.pos = v
  o.addToWorld(ph)

  # Walls (N, S, E, W)
  sh = ShBox(0.5*vec3(table_size.x+2*wall_width, wall_width, wall_height))
  o = OSimple(sh)
  o.color = Color4.black()
  o.pos = 0.5*vec3(0, table_size.y+wall_width, wall_height)
  o.addToWorld(ph)
  o = OSimple(sh)
  o.color = Color4.black()
  o.pos = 0.5*vec3(0, -table_size.y-wall_width, wall_height)
  o.addToWorld(ph)

  sh = ShBox(0.5*vec3(wall_width, table_size.y+2*wall_width, wall_height))
  o = OSimple(sh)
  o.color = Color4.black()
  o.pos = 0.5*vec3(table_size.x+wall_width, 0, wall_height)
  o.addToWorld(ph)
  o = OSimple(sh)
  o.color = Color4.black()
  o.pos = 0.5*vec3(-table_size.x-wall_width, 0, wall_height)
  o.addToWorld(ph)


def set_camera(di):
  """Set camera to look on the logo."""
  di.cam_mode = di.FIXED
  di.cam_target.cart = vec3(0.1,-0.2,0)
  di.cam_eye.spheric = spheric3(5,0.6,-1.0)


if __name__ == '__main__':
  from optparse import OptionParser

  parser = OptionParser(
      description="SimulOtter logo builder.",
      usage="%prog [OPTIONS] [FILE]")
  parser.add_option('-s', '--size', dest='size', metavar='X[,Y]',
      help="logo size (default: 320,320")
  parser.add_option('-a', '--antialias', dest='antialias', metavar='N',
      help="antialiasing level (default: 0, disabled)")
  parser.add_option('-b', '--background', dest='background', metavar='RGB',
      help="background color, as an RRGGBB hexa string (default: cccccc)")
  parser.set_defaults(
      size='320',
      antialias=0,
      background='cccccc',
      )
  opts, args = parser.parse_args()

  if ',' in opts.size:
    size = [ int(x) for x in opts.size.split(',',2) ]
  else:
    size = [ int(opts.size) ] * 2
  antialias = int(opts.antialias)
  if len(opts.background) != 6:
    parser.error("invalid background color format")
  bg_color = Color4( *(int(opts.background[i:i+2],16) for i in (0,1,2) ) )

  if len(args) == 0:
    fout = None
  elif len(args) == 1:
    fout = args[0]
  else:
    parser.error("extra arguments")

  Display.antialias = antialias
  ph = Physics()
  di = Display()
  di.physics = ph
  di.bg_color = bg_color

  build_logo(ph)
  set_camera(di)

  di.resize(size[0], size[1], False)
  if fout is not None:
    di.update()
    di.update()  # first frame may be broken on Linux
    di.screenshot(fout)
  else:
    di.run()

